### DISCUSSION TITLE
Display distance to the next intermediate stop rather than to the final destination

### DISCUSSION THREAD

**Discussion by Petak73** [organicmaps/organicmaps#8549] [Ideas]
Hi,

I think it would be great to have an option to display distance to the next intermediate stop (as opposed to the final destination, as it is now) in the info bar in the nav screen. Very useful when riding a bike and inserting a petrol station as an intermediate stop.

![Screenshot_20240624-075416](https://github.com/organicmaps/organicmaps/assets/173553515/13df3907-d434-41ed-a64f-e0864361960c)

**Comment by gpesquero** [comment]
As discussed in the Contributors' Telegram chat, we shall gather here information of how other apps handle the display of information to next route stop (display of route stops distance, remaining time and ETA -- Estimated Time of Arrival).

Related issues and PRs: https://github.com/organicmaps/organicmaps/issues/7762 https://github.com/organicmaps/organicmaps/pull/9071

**Reply by biodranik** [reply]
Displaying the time to the final destination when a stop is added does not make much sense in most cases. Because a stop implies that the car will be stopped, and something happens there for an unknown amount of time. So displaying ETA for the final destination won't help (a driver also remembers the initial ETA when the route was built).

Displaying the distance may be a bit more useful, but still confusing (e.g. show 12/27 km on the right).

A better approach would be to display only time and distance to the intermediate stop, but visually highlight it, so driver will understand that the total trip time/distance are different.

**Reply by ghost** [reply]
Defaulting to only displaying travel time, ETA and distance for the next stop when a stop is added is interesting. It would help keep the UI simple.

If someone would prefer to have travel time, ETA and distance displayed for the destination even when stops are added, maybe the default behaviour could be changed via settings?

PS: When a route is build (before tapping "Start") only travel time and distance are displayed. Maybe ETA for the destination could be added as well to help users get a sense of the entire route (particularly if they've added stops and will not have that info readily available once they tap "Start")?

**Comment by gpesquero** [comment]
**OsmAnd**

* OsmAnd allows setting multiple route stops.
* It displays info to final destination (distance + time OR ETA) in a mini panel located at the top-right of the navigation screen.
* When stops are set, it displays in this same mini panel the (distance+time/ETA) for both final and next stop.
* If clicking on the remaining distance, a route panel is displayed with information of the whole route.
* If clicking on the remaining time or ETA, display switched between remaining time or ETA.

![OsmAnd_low](https://github.com/user-attachments/assets/61828e1f-877a-40f1-9976-6fe67753e2f1)

**Reply by biodranik** [reply]
OSMand targets a very niche audience. Let's focus on more mainstream driving app examples.

**Reply by Petak73** [reply]
OSMand is one of the most popular/downloaded nav apps in the Play Store. Let's not discard it as "niche", please :-)

**Reply by euf** [reply]
When there's an intermediate stop on the route, Waze shows ETA for it, not for the final destination.

**Comment by gpesquero** [comment]
**Waze**

 * Waze only allows to set a single stop.
 * It normally displays info to final destination in a bottom panel (as OM), showing all 3 distance+time+ETA.
 * When an intermediate stop is set, the info to the final destination is replaced with the info to the stop.
 * The color of the route is slightly changed when a stop is set.
 * It shows no info to the final destination.
 * You can only know the ETA to the final destination if you click on the bottom panel, and then a complete route info panel is displayed.

With **no** route stop:
![Waze_dest_low](https://github.com/user-attachments/assets/7462921f-e570-44a5-8410-7e37416358fa)

With route stop:
![Waze_stop_low](https://github.com/user-attachments/assets/23d50475-682e-485e-8f2e-251d5aaf1bcb)

**Comment by gpesquero** [comment]
**Google Maps**

 * It works the same as Waze, but it allows multiple intermediate stops and it does not change the route color, though.
 * It does not show anywhere the ETA, time nor distance to final destination.

**Comment by gpesquero** [comment]
**Apple Maps**

 * It allows setting up multiple stops.
 * If no stop is set, it displays ETA, remaining time and distance to destination. The label "Llegada" (Finish in Spanish) is displayed under the ETA value:
 
![apple_finish](https://github.com/user-attachments/assets/51bc3de9-36be-4f9f-86c1-9736b5990e0c)

 * If a stop is set, it displays the information to the intermediate stop and the label "Parada" (Stop in Spanish) under the ETA value:

![apple_stop](https://github.com/user-attachments/assets/c746fd2b-a436-4f39-8732-05597286b365)

**Reply by biodranik** [reply]
Nice! An elegant solution.

**Reply by ghost** [reply]
I agree.

**Comment by gpesquero** [comment]
**Mapy cz (Android)**

 * It allows setting up multiple stops.
 * If no stop is set, it displays ETA, remaining time and distance to destination. It always displays the label "Arrival" under the ETA value.
 * If an intermediate stop is set, it still displays the info to the final destination.
 * There's the possibility though to enable in the "Navigation Settings" of the app the display of the info to next stop close to the navigation arrow:
![Mapy cz](https://github.com/user-attachments/assets/4821c66a-12c5-412c-9726-86eaaf2b89a6)

**Comment by gpesquero** [comment]
**Sygic (Android)**

 * It always displays the ETA, remaining time and distance to final destination, even if a "waypoint" is set.
 * There seems to be no way to know the info to the next stop, unless you get into the menus to get the route information.

![Sygic](https://github.com/user-attachments/assets/6ce200d8-535e-44f6-9809-1e38214d4977)

**Comment by gpesquero** [comment]
@biodranik, @organicmaps/design: I think that we've already gathered enough information of the implementation of the next stop info display on other apps. Should we start the discussion here of the possible implementations in OM?

**Comment by Rob-from-VI** [comment]
This is a feature I definitely would find useful.  Was actually looking for a place to post it. I have several use cases.

Sometimes, a stop is added to direct a route via a particular intermediate spot, without any intention of stopping, so in this case total time, distance and ETA would be more useful

In other cases if I actually intend to stop, I might like to know how long I can stop without exceeding my final arrival deadline.   In this case both items are needed.

In yet other cases I might want to stop for an extended period, but leave my final destination preprogrammed for when I leave the stop.  In this case only info related to the stop is needed.

So when a stop is set I definitely **need the option** somehow to display my choice of data to the stop or data to final destination.

So, would it be possible to have the interface simply be when I touch the final arrival time display it switches to first stop data, and touching it again brings back final destination?   

I find on Android Auto the text is too large, and if it was a bit smaller we could have all three items ie time, distance, and eta.   If all three is deemed to mch then keep as is ie time and distance an lose eta.

**Comment by gpesquero** [comment]
Based on the solution of many apps that display the info to the next stop in the same panel as the final destination, I've made some tests on a possible solution that shows an icon on the bottom panel that enables the switching of the info display between next stop or final destination:

![imagen1](https://github.com/user-attachments/assets/1c444a4d-f1a2-4f1f-aa9e-44c090c88dd1)

![imagen2](https://github.com/user-attachments/assets/73ccedc2-cfc8-4d81-a093-f9c44ccb5df8)

By clicking on the new icon, the user can switch between next stop info or final destination info.

**Comment by gpesquero** [comment]
The info display might get too "overcrowded" if the info to display has large time and distance values, as in this example:

![imagen3](https://github.com/user-attachments/assets/4ac303a9-2b80-4213-af17-b9dbf5e54446)

![imagen4](https://github.com/user-attachments/assets/32fbf0b1-ed46-49aa-9802-964aa2e083ce)

**Comment by Rob-from-VI** [comment]
I like it.  So glad someone is taking it up.  I understand we need an icon to show what is being displayed.

There appears to be some spare space between the speed limit and the time, which could be balanced out.  Basically you are only adding about 3 characters width.  

 Also, just noticing for remaining distance the KM appears behind the number in your picture, as well as undeneath it.  We only need it underneath, and this would make up for it.

Actually on my phone screen this info rectangle extends full width of the screen and thus lots of room.  By the way my phone does not have the additional KM to the right of the distance

**Reply by gpesquero** [reply]
> Also, just noticing for remaining distance the KM appears behind the number in your picture, as well as undeneath it. We only need it underneath, and this would make up for it.

You're right @Rob-from-VI : I hacked the code to simulate the display of large distances, and I incorrectly added the units there. I'm going to fix it and upload updated screenshots.

**Comment by gpesquero** [comment]
Here are the "fixed" screenshots of the display with large numbers (it does not seem too "crowed" now): 

![imagen6](https://github.com/user-attachments/assets/27f54e61-1ecd-442e-8f51-a3ac13b411de)

![imagen7](https://github.com/user-attachments/assets/ee70cdd0-0457-4d3c-a8a8-2f7fb2526a59)

The spacing and location of the different fields can be improved for a better look...

**Comment by gpesquero** [comment]
One other possible implementation would be to add the next stop/final destination icon in the hidden bottom bar:

![imagen8](https://github.com/user-attachments/assets/6108092c-d0ed-4eb0-89c0-a05a78ae1e42)

This would make the main bar cleaner, but on the contrary it would be less intuitive. The use may not know exactly which distance/time info is being displayed (the one to the next stop, or to the final destination), unless the hidden bottom bar is shown.

**Comment by Rob-from-VI** [comment]
I think it would be nice to see the icon at all times, so we can tell what is being displayed, and as you say easier to find.

I also think having the Icon at the very beginning or at the end of the line (behind the distance kms) rather than middle would be better

**Reply by gpesquero** [reply]
> I also think having the Icon at the very beginning or at the end of the line (behind the distance kms) rather than middle would be better.

@Rob-from-VI: Here's a possible implementation showing the icon at the end of the navigation bar:

**No intermediate stop:**
![picture1](https://github.com/user-attachments/assets/cb9109d9-fcfe-45da-83ee-87e37fbbe609)

**Show info to next stop:**
![picture2](https://github.com/user-attachments/assets/8a6809c3-3b57-4ca3-be4d-a214237290db)

**Show info to final destination:**
![picture3](https://github.com/user-attachments/assets/2ce9c723-cdca-4679-a946-8a4befd0090e)

**Comment by Rob-from-VI** [comment]
So I think we really only need version 2 and 3.  In case there is no intermediate stop, it would simply display the flag for final destination

I think the default should be next stop if a stop is included.

**Reply by gpesquero** [reply]
> So I think we really only need version 2 and 3. In case there is no intermediate stop, it would simply display the flag for final destination

You mean **always** displaying the flag for final destination, even if no stop is set? We would be always misusing that space and leaving less room for the texts.



> I think the default should be next stop if a stop is included.

I agree that the default should be to display the info to the next stop.

**Comment by gpesquero** [comment]
With the icon set at the right of the nav bar, we're also shifting the texts to the left, which may not be the optimal visual approach.

As we also have to support low resolution devices, I've taken some screenshots with the Android simulation set to 480 pixel width:

The display is already quite saturated with no icon:
![picture1](https://github.com/user-attachments/assets/89812e3c-3732-433f-aae5-7f5a6f2bb5d1)

And the icons are hidden partially if a stop is set: 
![picture2](https://github.com/user-attachments/assets/10d58078-e3fc-4f7b-bc44-935dff393eca)
![picture3](https://github.com/user-attachments/assets/04c63a6d-e605-44ef-aa27-6d4326c3b753)

I could try to improve the spacing handling between the different items, but I think that the navigation bar is going to be quite saturated with this approach; the numbers are not going to be readable as they are too close to each other.

The fact that we're displaying the current speed **and** the speed limit in the format XXX / YYY does not really help either. As this is a temporary solution, maybe we shall wait for [PR 7005](https://github.com/organicmaps/organicmaps/pull/7005) to be merged so we have more space to include the next stop icon (somewhere).

**Reply by Rob-from-VI** [reply]
If you want to pursue the single panel solution, how about dropping the  min or at least the in from 3h25min.   If the first digit is hours I think it is obvious what the next two digits are.   This would give enough space.   

I would also be quite happy with the separate panel showing stop distance and eta underneath next turn, so either way lets somebody decide, and get on with it!!!

**Reply by ghost** [reply]
Waiting for #7005 to be merged could be interesting.. Then something like this could be a possible solution:

![v2 Destination info - Organic Maps](https://github.com/user-attachments/assets/134f0774-e729-4762-a23f-ba2a6ddcb52c)
![v2 Next stop info - Organic Maps](https://github.com/user-attachments/assets/764ad34c-d582-45c2-9f50-9770b4356f15)

Or something like this:

![v3 Destination info - Organic Maps](https://github.com/user-attachments/assets/2f14af94-1af6-45ff-aa83-f4c46270fe96)
![v3 Next stop info - Organic Maps](https://github.com/user-attachments/assets/52a52607-6f77-4abb-bd92-3047a2dd46e3)

Tapping on the light grey area could switch between displaying info for the next stop and info for the destination.

**Reply by Rob-from-VI** [reply]
I like the idea of tapping anywhere in the grey area.  Remains to be seen how this can be displayed on the Android Auto or Carplay, which is where it would be most important!!!

But please don't lose the Actual speed/speed limit.  That is VERY useful.

**Comment by gpesquero** [comment]
And many people don't like this first initial approach proposed in https://github.com/organicmaps/organicmaps/pull/9071, but I still find it quite convenient:

![361234119-55fcfac4-37bd-4c0e-894f-aa9bded8a0e5](https://github.com/user-attachments/assets/453065f3-df90-477b-853e-4cd8168ec9e6)

(Stop info panel design courtesy of @tobiaalberti)

**Reply by Rob-from-VI** [reply]
Sure, I think this is OK too, no need to push any buttons, see everything all at once

Except possibly touch once to see ETA time instead of time remaining
